### ISSUE TITLE
use Flair with skweak

### ISSUE THREAD

**Issue by Ihebzayen** [NorskRegnesentral/skweak#54]
hello ,
is here anyone who tried to implement another model/framework other than spacy (ner) as a labeling function.
i tried to work with flair but didnt work.
can anyone help me and thanks in advance .

**Comment by neilmizzi** [comment]
> hello , is here anyone who tried to implement another model/framework other than spacy (ner) as a labeling function. i tried to work with flair but didnt work. can anyone help me and thanks in advance .

I am also having this issue at the moment! So far, I've been able to use external models which can be fit into the SpaCy framework: https://spacy.io/universe/category/models

I'm currently trying to use Flair using a myriad of ways (the SpaCy-wrap tool, changing to the SpaCy tokenizer), but so far have not had any success. I will update if I manage to do this!

In the meantime, having the option to integrate non-SpaCy-based models more easily would definitely be appreciated. I will see how far I get with trying to get Flair to work and keep this thread updated.

**Comment by plison** [comment]
Yes, I agree that using character spans instead of token-level spans would make `skweak` less spacy-dependent and provide more flexibility. But it would mean rewriting a lot of the code, since right now, the results of the labelling functions are stored as `Span` objects, which require token-level indexing. So it's definitely something that would be worth looking into, but it's not the pipeline at the moment.

**Comment by neilmizzi** [comment]
Thank you for your input @plison! I can understand that this may not be as straightforward to implement.

At the moment, I have been able to figure out a workaround to get Flair working using the `FunctionAnnotator`, by using the function `doc.char_span` to manually set the entities such that they are supported (and then tokenized) using the existing `doc` object. You can see an example below.

This workaround should in theory work for any model which provides character spans for the retrieved entities, however I have only tested it with Flair so far, and haven't had any major issues.

```py
from skweak import heuristics

from flair.data import Sentence
from flair.models import SequenceTagger

...

flair_classifier = SequenceTagger.load("flair/ner-english-large")

    def flair_annotator(doc):
        sentence = Sentence(doc.text)
        flair_classifier.predict(sentence)

        spans = []
        for entity in sentence.get_spans('ner'):
            spans.append(doc.char_span(entity.start_position, entity.end_position, entity.labels[0].value))

        for token in spans:
            if token:   # exclude NoneType Entities
                yield token.start, token.end, token.label

...

# declare and set Flair annotations
flair_annotator = heuristics.FunctionAnnotator("flair_annotator", flair_annotator)

docs = list(flair_annotator.pipe(docs))
...
```

**Comment by plison** [comment]
Thanks, that's very useful indeed!
